Cannot find module './lib/cookies'

Did an update:
```
This project is already at Meteor 1.4.4.2, the latest release.

Changes to your project's package version selections from updating package versions:

ostrio:cookies             upgraded from 2.2.0 to 2.2.1
ostrio:files               upgraded from 1.7.13 to 1.7.15
react-meteor-data          upgraded from 0.2.9 to 0.2.11
reywood:publish-composite  upgraded from 1.5.1 to 1.5.2
```

node Version: 4.8.3
npm Version: 4.6.1

--> Error:
```
Error: Cannot find module './lib/cookies'
    at require (packages/modules-runtime.js:123:19)
    at meteorInstall.node_modules.request.index.js (packages/modules.js:9409:29)
    at fileEvaluate (packages/modules-runtime.js:197:9)
    at require (packages/modules-runtime.js:120:16)
    at meteorInstall.node_modules.meteor.ostrio:files.files.coffee.js (packages/ostrio_files/files.coffee:10:14)
    at fileEvaluate (packages/modules-runtime.js:197:9)
    at require (packages/modules-runtime.js:120:16)
    at /Users/patrick/dev/orderlion/orderlion_react/app/.meteor/local/build/programs/server/packages/ostrio_files.js:4839:15
    at /Users/patrick/dev/orderlion/orderlion_react/app/.meteor/local/build/programs/server/packages/ostrio_files.js:4850:3
    at /Users/patrick/dev/orderlion/orderlion_react/app/.meteor/local/build/programs/server/boot.js:303:34
```
Any ideas on this?

Can't start my app anymore! :(

best, P

**EDIT:**
I rolled back to 1.7.13, still same error. :(

**EDIT 2:**
Had to completely remove your package for now. I tried deleting `.meteor/local` and `node_modules`, thus re-downloading and installing all packages and stuff. same problem. I hope you guys can help ... I am a bit lost what exactly completely broke my app. It is definitely cause by your package (or maybe your "cookie" package) as my app starts up again now.

--- a/.versions
+++ b/.versions
@@ -36,7 +36,7 @@ npm-mongo@2.2.24
 observe-sequence@1.0.16
 ordered-dict@1.0.9
 ostrio:cookies@2.2.1
-ostrio:files@1.7.15
+ostrio:files@1.7.16
 promise@0.8.8
 random@1.0.10
 reactive-var@1.0.11

--- a/docs/aws-s3-integration.md
+++ b/docs/aws-s3-integration.md
@@ -150,25 +150,23 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
       }
 
       if (path) {
-        //pipe the request from S3 through us and direct into the http socket
-        //need to set the header so the file shows up properly
+        // If file is successfully moved to AWS:S3
+        // We will pipe request to AWS:S3
+        // So, original link will stay always secure
 
-        //this is needed if you want to view the files in the browser
-        //set the content type and lenth to what we recorded
-        http.response.setHeader('content-type', fileRef.type || 'application/pdf');
-        http.response.setHeader('content-length', fileRef.size || 0);
-
-
-        //get the file/key and pipe it into out http response
-        s3.getObject({
+        // To force ?play and ?download parameters
+        // and to keep original file name, content-type,
+        // content-disposition, chunked "streaming" and cache-control
+        // we're using low-level .serve() method
+        const opts = {
           Bucket: s3Conf.bucket,
-          Key: path
-        }).createReadStream().on('error', (err) => {
-          bound(() => {
-            console.error(err);
-          });
-        }).pipe(http.response);
+          Key: path,
+        };
 
+        if (http.request.headers.range) {
+          opts.Range = http.request.headers.range;
+        }
+        this.serve(http, fileRef, fileRef.versions[version], version, client.getObject(opts).createReadStream());
         return true;
       }
       // While file is not yet uploaded to AWS:S3

--- a/docs/aws-s3-integration.md
+++ b/docs/aws-s3-integration.md
@@ -58,6 +58,7 @@ import { Meteor } from 'meteor/meteor';
 import { _ } from 'meteor/underscore';
 import { Random } from 'meteor/random';
 import { FilesCollection } from 'meteor/ostrio:files';
+import stream from 'stream';
 
 import S3 from 'aws-sdk/clients/s3'; // http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html
 // See fs-extra and graceful-fs NPM packages
@@ -160,14 +161,28 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
         // we're using low-level .serve() method
         const opts = {
           Bucket: s3Conf.bucket,
-          Key: path,
+          Key: path
         };
 
         if (http.request.headers.range) {
           opts.Range = http.request.headers.range;
         }
-        this.serve(http, fileRef, fileRef.versions[version], version, client.getObject(opts).createReadStream());
-        return true;
+
+        const fileColl = this;
+        client.getObject(opts, function (error, data) {
+          if (error) {
+            console.error(error);
+            http.response.end();
+          } else {
+            if (http.request.headers.range && this.httpResponse.headers['content-range']) {
+              http.request.headers.range = this.httpResponse.headers['content-range'].split('/')[0].replace('bytes ', 'bytes=');
+            }
+
+            const dataStream = new stream.PassThrough();
+            fileColl.serve(http, fileRef, fileRef.versions[version], version, dataStream);
+            dataStream.end(this.data.Body);
+          }
+        });
       }
       // While file is not yet uploaded to AWS:S3
       // It will be served file from FS

--- a/docs/aws-s3-integration.md
+++ b/docs/aws-s3-integration.md
@@ -83,7 +83,11 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
     secretAccessKey: s3Conf.secret,
     accessKeyId: s3Conf.key,
     region: s3Conf.region,
-    sslEnabled: true
+    // sslEnabled: true, // optional
+    httpOptions: {
+      timeout: 6000,
+      agent: false
+    }
   });
 
   // Declare the Meteor file collection on the Server
@@ -109,8 +113,8 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
         // Key is the file name we are creating on AWS:S3, so it will be like files/XXXXXXXXXXXXXXXXX-original.XXXX
         // Body is the file stream we are sending to AWS
         s3.putObject({
-          ServerSideEncryption: 'AES256',
-          StorageClass: 'STANDARD_IA',
+          // ServerSideEncryption: 'AES256', // Optional
+          StorageClass: 'STANDARD',
           Bucket: s3Conf.bucket,
           Key: filePath,
           Body: fs.createReadStream(vRef.path),
@@ -165,16 +169,32 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
         };
 
         if (http.request.headers.range) {
-          opts.Range = http.request.headers.range;
+          const vRef  = fileRef.versions[version];
+          let range   = _.clone(http.request.headers.range);
+          const array = range.split(/bytes=([0-9]*)-([0-9]*)/);
+          const start = parseInt(array[1]);
+          let end     = parseInt(array[2]);
+          if (isNaN(end)) {
+            // Request data from AWS:S3 by small chunks
+            end       = (start + this.chunkSize) - 1;
+            if (end >= vRef.size) {
+              end     = vRef.size - 1;
+            }
+          }
+          opts.Range   = `bytes=${start}-${end}`;
+          http.request.headers.range = `bytes=${start}-${end}`;
         }
 
         const fileColl = this;
-        client.getObject(opts, function (error, data) {
+        s3.getObject(opts, function (error) {
           if (error) {
             console.error(error);
-            http.response.end();
+            if (!http.response.finished) {
+              http.response.end();
+            }
           } else {
             if (http.request.headers.range && this.httpResponse.headers['content-range']) {
+              // Set proper range header in according to what is returned from AWS:S3
               http.request.headers.range = this.httpResponse.headers['content-range'].split('/')[0].replace('bytes ', 'bytes=');
             }
 
@@ -183,6 +203,8 @@ if (s3Conf && s3Conf.key && s3Conf.secret && s3Conf.bucket && s3Conf.region) {
             dataStream.end(this.data.Body);
           }
         });
+
+        return true;
       }
       // While file is not yet uploaded to AWS:S3
       // It will be served file from FS

--- a/files.coffee
+++ b/files.coffee
@@ -2384,9 +2384,9 @@ class FilesCollection
       responseType = '200'
 
     streamErrorHandler = (error) ->
-      http.response.writeHead 500
-      http.response.end error.toString()
       console.error "[FilesCollection] [serve(#{vRef.path}, #{version})] [500]", error if self.debug
+      if !http.response.finished
+        http.response.end error.toString()
       return
 
     headers = if _.isFunction(self.responseHeaders) then self.responseHeaders(responseType, fileRef, vRef, version) else self.responseHeaders

--- a/files.coffee
+++ b/files.coffee
@@ -5,12 +5,12 @@ if Meteor.isServer
   ###
   @summary Require NPM packages
   ###
-  fs       = require 'fs-extra'
-  events   = require 'events'
-  request  = require 'request'
-  Throttle = require 'throttle'
-  fileType = require 'file-type'
-  nodePath = require 'path'
+  fs       = Npm.require 'fs-extra'
+  events   = Npm.require 'events'
+  request  = Npm.require 'request'
+  Throttle = Npm.require 'throttle'
+  fileType = Npm.require 'file-type'
+  nodePath = Npm.require 'path'
 
   ###
   @var {Object} bound - Meteor.bindEnvironment (Fiber wrapper)
@@ -758,10 +758,12 @@ class FilesCollection
           console.warn '[FilesCollection._checkAccess] WARN: Access denied!' if self.debug
           if http
             text = 'Access denied!'
-            http.response.writeHead rc,
-              'Content-Length': text.length
-              'Content-Type':   'text/plain'
-            http.response.end text
+            if !http.response.headersSent
+              http.response.writeHead rc,
+                'Content-Length': text.length
+                'Content-Type':   'text/plain'
+            if !http.response.finished
+              http.response.end text
           return false
       else
         return true
@@ -782,8 +784,10 @@ class FilesCollection
 
             handleError = (error) ->
               console.warn "[FilesCollection] [Upload] [HTTP] Exception:", error
-              response.writeHead 500
-              response.end JSON.stringify {error}
+              if !response.headersSent
+                response.writeHead 500
+              if !response.finished
+                response.end JSON.stringify {error}
               return
 
             body = ''
@@ -820,19 +824,29 @@ class FilesCollection
 
                   if opts.eof
                     self._handleUpload result, opts, ->
-                      response.writeHead 200
+                      if !response.headersSent
+                        response.writeHead 200
                       result.file.meta = fixJSONStringify result.file.meta if result?.file?.meta
-                      response.end JSON.stringify result
+                      if !response.finished
+                        response.end JSON.stringify result
                       return
                     return
                   else
                     self.emit '_handleUpload', result, opts, NOOP
 
-                  response.writeHead 204
-                  response.end()
+                  if !response.headersSent
+                    response.writeHead 204
+                  if !response.finished
+                    response.end()
 
                 else
-                  opts           = JSON.parse body
+                  try
+                    opts         = JSON.parse body
+                  catch e
+                    console.error 'Can\'t parse incoming JSON from Client on [.insert() | upload], something went wrong!'
+                    console.error e
+                    opts         = file: {}
+
                   opts.___s      = true
                   console.info "[FilesCollection] [File Start HTTP] #{opts.file.name} - #{opts.fileId}" if self.debug
                   opts.file.meta = fixJSONParse opts.file.meta if opts?.file?.meta
@@ -845,14 +859,18 @@ class FilesCollection
                   self._createStream result._id, result.path, _.omit(opts, '___s')
 
                   if opts.returnMeta
-                    response.writeHead 200
-                    response.end JSON.stringify {
-                      uploadRoute: "#{self.downloadRoute}/#{self.collectionName}/__upload"
-                      file: result
-                    }
+                    if !response.headersSent
+                      response.writeHead 200
+                    if !response.finished
+                      response.end JSON.stringify {
+                        uploadRoute: "#{self.downloadRoute}/#{self.collectionName}/__upload"
+                        file: result
+                      }
                   else
-                    response.writeHead 204
-                    response.end()
+                    if !response.headersSent
+                      response.writeHead 204
+                    if !response.finished
+                      response.end()
               catch error
                 handleError error
               return
@@ -1857,7 +1875,11 @@ class FilesCollection
               'x-fileId':     opts.fileId
               'content-type': 'text/plain'
           }, (error, result) ->
-            result      = JSON.parse result?.content or {}
+            try
+              result    = JSON.parse result?.content or {}
+            catch e
+              console.warn 'Something went wrong! [sendEOF] method doesn\'t returned JSON! Looks like you\'re on Cordova app or behind proxy, switching to DDP transport is recommended.'
+              result    = {}
             result.meta = fixJSONParse result.meta if result?.meta
             self.emitEvent 'end', [error, result]
             return
@@ -2271,10 +2293,13 @@ class FilesCollection
   _404: if Meteor.isServer then (http) ->
     console.warn "[FilesCollection] [download(#{http.request.originalUrl})] [_404] File not found" if @debug
     text = 'File Not Found :('
-    http.response.writeHead 404,
-      'Content-Length': text.length
-      'Content-Type':   'text/plain'
-    http.response.end text
+
+    if !http.response.headersSent
+      http.response.writeHead 404,
+        'Content-Length': text.length
+        'Content-Type':   'text/plain'
+    if !http.response.finished
+      http.response.end text
     return
   else undefined
 
@@ -2348,10 +2373,11 @@ class FilesCollection
     else
       dispositionType = 'inline; '
 
-    dispositionName     = "filename=\"#{encodeURIComponent(fileRef.name)}\"; filename=*UTF-8\"#{encodeURIComponent(fileRef.name)}\"; "
-    dispositionEncoding = 'charset=utf-8'
+    dispositionName     = "filename=\"#{encodeURI(fileRef.name)}\"; filename*=UTF-8''#{encodeURI(fileRef.name)}; "
+    dispositionEncoding = 'charset=UTF-8'
 
-    http.response.setHeader 'Content-Disposition', dispositionType + dispositionName + dispositionEncoding
+    if !http.response.headersSent
+      http.response.setHeader 'Content-Disposition', dispositionType + dispositionName + dispositionEncoding
 
     if http.request.headers.range and not force200
       partiral = true
@@ -2392,53 +2418,101 @@ class FilesCollection
     headers = if _.isFunction(self.responseHeaders) then self.responseHeaders(responseType, fileRef, vRef, version) else self.responseHeaders
 
     unless headers['Cache-Control']
-      http.response.setHeader 'Cache-Control', self.cacheControl
+      if !http.response.headersSent
+        http.response.setHeader 'Cache-Control', self.cacheControl
 
     for key, value of headers
-      http.response.setHeader key, value
+      if !http.response.headersSent
+        http.response.setHeader key, value
 
     switch responseType
       when '400'
         console.warn "[FilesCollection] [serve(#{vRef.path}, #{version})] [400] Content-Length mismatch!" if self.debug
         text = 'Content-Length mismatch!'
-        http.response.writeHead 400,
-          'Content-Type':   'text/plain'
-          'Content-Length': text.length
-        http.response.end text
+
+        if !http.response.headersSent
+          http.response.writeHead 400,
+            'Content-Type':   'text/plain'
+            'Content-Length': text.length
+        if !http.response.finished
+          http.response.end text
         break
       when '404'
         return self._404 http
         break
       when '416'
         console.warn "[FilesCollection] [serve(#{vRef.path}, #{version})] [416] Content-Range is not specified!" if self.debug
-        http.response.writeHead 416
-        http.response.end()
+        if !http.response.headersSent
+          http.response.writeHead 416
+        if !http.response.finished
+          http.response.end()
         break
       when '200'
         console.info "[FilesCollection] [serve(#{vRef.path}, #{version})] [200]" if self.debug
         stream = readableStream or fs.createReadStream vRef.path
-        http.response.writeHead 200 if readableStream
+        if !http.response.headersSent
+          http.response.writeHead 200 if readableStream
+
+        http.response.on 'close', ->
+          stream.abort?()
+          stream.end?()
+          return
+
+        http.request.on 'abort', ->
+          stream.abort?()
+          stream.end?()
+          return
+
         stream.on('open', ->
-          http.response.writeHead 200
+          if !http.response.headersSent
+            http.response.writeHead 200
+          return
+        ).on('abort', ->
+          if !http.response.finished
+            http.response.end()
+          if !http.request.aborted
+            http.request.abort()
           return
         ).on('error', streamErrorHandler
         ).on 'end', ->
-          http.response.end()
+          if !http.response.finished
+            http.response.end()
           return
         stream.pipe new Throttle {bps: self.throttle, chunksize: self.chunkSize} if self.throttle
         stream.pipe http.response
         break
       when '206'
         console.info "[FilesCollection] [serve(#{vRef.path}, #{version})] [206]" if self.debug
-        http.response.setHeader 'Content-Range', "bytes #{reqRange.start}-#{reqRange.end}/#{vRef.size}"
+        if !http.response.headersSent
+          http.response.setHeader 'Content-Range', "bytes #{reqRange.start}-#{reqRange.end}/#{vRef.size}"
         stream = readableStream or fs.createReadStream vRef.path, {start: reqRange.start, end: reqRange.end}
-        http.response.writeHead 206 if readableStream
+        if !http.response.headersSent
+          http.response.writeHead 206 if readableStream
+
+        http.response.on 'close', ->
+          stream.abort?()
+          stream.end?()
+          return
+
+        http.request.on 'abort', ->
+          stream.abort?()
+          stream.end?()
+          return
+
         stream.on('open', ->
-          http.response.writeHead 206
+          if !http.response.headersSent
+            http.response.writeHead 206
+          return
+        ).on('abort', ->
+          if !http.response.finished
+            http.response.end()
+          if !http.request.aborted
+            http.request.abort()
           return
         ).on('error', streamErrorHandler
         ).on 'end', ->
-          http.response.end()
+          if !http.response.finished
+            http.response.end()
           return
         stream.pipe new Throttle {bps: self.throttle, chunksize: self.chunkSize} if self.throttle
         stream.pipe http.response

--- a/package.js
+++ b/package.js
@@ -1,6 +1,6 @@
 Package.describe({
   name: 'ostrio:files',
-  version: '1.7.15',
+  version: '1.7.16',
   summary: 'File upload via DDP/HTTP to server FS, AWS, GridFS, DropBox, Google Drive or other 3rd party storage',
   git: 'https://github.com/VeliovGroup/Meteor-Files',
   documentation: 'README.md'

